src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs

```diff
@@ -52,59 +56,113 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
+                        headerBuffer.CopyTo(_headerBuffer.AsMemory(HeaderLength - _contentRemaining));
+
+                        _contentRemaining -= headerBuffer.Length;
+                        if (_contentRemaining > 0)
+                        {
+                            return read;
+                        }
+
+                        headerBuffer = _headerBuffer;
+                    }
+
+                    var length = (int)BinaryPrimitives.ReadUInt32BigEndian(headerBuffer.Span.Slice(1));
+
+                    _contentRemaining = length;
+                    // If there is no content then state is reset to ready.
+                    _state = _contentRemaining > 0 ? ResponseState.Content : ResponseState.Ready;
+                    return read;
                 }
-
-                var success = await TryReadDataAsync(_inner, headerBuffer, cancellationToken).ConfigureAwait(false);
-                if (!success)
+            case ResponseState.Content:
                 {
-                    return 0;
+                    if (data.Length >= _contentRemaining)
+                    {
+                        data = data.Slice(0, _contentRemaining);
+                    }
+
+                    var read = await StreamHelpers.ReadAsync(_inner, data, cancellationToken).ConfigureAwait(false);
+                    _contentRemaining -= read;
+                    if (_contentRemaining == 0)
+                    {
+                        _state = ResponseState.Ready;
+                    }
+
+                    return read;
                 }
-
-                var compressed = headerBuffer.Span[0];
-                var length = (int)BinaryPrimitives.ReadUInt32BigEndian(headerBuffer.Span.Slice(1));
-
-                var isTrailer = IsBitSet(compressed, pos: 7);
-                if (isTrailer)
+            case ResponseState.Trailer:
                 {
+                    Debug.Assert(headerBuffer.Length > 0);
+
+                    // The trailer needs to be completely read before returning 0 to the caller.
+                    // Ensure buffer is large enough to contain the trailer header.
+                    if (headerBuffer.Length < HeaderLength)
+                    {
+                        var newBuffer = new byte[HeaderLength];
+                        headerBuffer.CopyTo(newBuffer);
+                        var success = await TryReadDataAsync(_inner, newBuffer.AsMemory(headerBuffer.Length), cancellationToken).ConfigureAwait(false);
+                        if (!success)
+                        {
+                            return 0;
+                        }
+                        headerBuffer = newBuffer;
+                    }
+                    var length = (int)BinaryPrimitives.ReadUInt32BigEndian(headerBuffer.Span.Slice(1));
+
                     await ReadTrailersAsync(length, data, cancellationToken).ConfigureAwait(false);
                     return 0;
```

Comment:
What's a trailer and why do we always return "0 bytes read" for them?

Reply:
My read is that they're not pard of the content payload, and thus we're not reporting them as part of the Stream content; likewise, once we've seen a trailer we know we're at the EOF, hence: return 0 to signal that. But I'm very interested in whether my interpretation was right!

Reply:
You're correct. Typically, gRPC uses HTTP/2's built-in trailing header functionality for the "grpc-status" and "grpc-message" trailers, but [gRPC Web encodes this in the response body](https://fuchsia.googlesource.com/third_party/grpc/+/refs/tags/v1.6.5/doc/PROTOCOL-WEB.md#protocol-differences-vs-grpc-over-http2). Waiting for the response body to complete before sending the status allows services to report errors that occur after they start streaming a response.

[HTTP/1.1 also technically supports trailing headers for chunked responses](https://httpwg.org/specs/rfc9112.html#chunked.trailer.section), but support is spotty. I assume that's why gRPC Web doesn't rely on it.

Reply:
> My read is that they're not pard of the content payload, and thus we're not reporting them as part of the Stream content; likewise, once we've seen a trailer we know we're at the EOF, hence: return 0 to signal that. But I'm very interested in whether my interpretation was right!

Correct.